src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/MarkDuplicatesSparkIntegrationTest.java

```diff
@@ -175,6 +174,308 @@ public void testMarkDuplicatesSparkIntegrationTestLocal(
         }
     }
 
+    @Test( dataProvider = "md")
+    public void testMarkDuplicatesSparkMarkingOpticalDuplicatesWithTagging(
+            final File input, final long totalExpected, final long dupsExpected,
+            Map<String, List<String>> metricsExpected) throws IOException {
+
+        ArgumentsBuilder args = new ArgumentsBuilder();
+        args.add("--" + StandardArgumentDefinitions.INPUT_LONG_NAME);
+        args.add(input.getPath());
+        args.add("--" + StandardArgumentDefinitions.OUTPUT_LONG_NAME);
+
+        File outputFile = createTempFile("markdups", ".bam");
+        outputFile.delete();
+        args.add(outputFile.getAbsolutePath());
+
+        args.add("--" + StandardArgumentDefinitions.METRICS_FILE_LONG_NAME);
+        File metricsFile = createTempFile("markdups_metrics", ".txt");
+        args.add(metricsFile.getAbsolutePath());
+
+        args.add("--" + MarkDuplicatesSparkArgumentCollection.DUPLICATE_TAGGING_POLICY_LONG_NAME);
+        args.add(MarkDuplicates.DuplicateTaggingPolicy.OpticalOnly);
+
+        runCommandLine(args.getArgsArray());
+
+        Assert.assertTrue(outputFile.exists(), "Can't find expected MarkDuplicates output file at " + outputFile.getAbsolutePath());
+
+        Set<String> opticalDuplicateNamesObserved = new HashSet<>();
+
+        final MetricsFile<GATKDuplicationMetrics, Comparable<?>> metricsOutput = new MetricsFile<>();
+        try {
+            metricsOutput.read(new FileReader(metricsFile));
+        } catch (final FileNotFoundException ex) {
+            System.err.println("Metrics file not found: " + ex);
+        }
+        final List<GATKDuplicationMetrics> nonEmptyMetrics = metricsOutput.getMetrics().stream().filter(
```

Comment:
this is a big block of complicated duplicated code, could you extract it to a method instead?